### PR TITLE
Infra/eslint v2

### PR DESCRIPTION
**Pull request by M-i-k-e-l**
## Description
Handling more types of imports\requires (for instance importing *).
Handling renamed components.
Our fix is not working on all the examples.
Refactor to have more and simpler to understand utils.

It's a big one, don't get too stressed over it. Try to see what you can use later from your new rule point-of-view (you'll need to edit it and improve using these utils).
Perhaps try to see if I missed something in the tests etc.

## Changelog
Big refactor, support more import\require types (for instance importing *), handling renamed components, improve fix logic.

### CODE REVIEW

File: eslint-rules/lib/utils/importUtils.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
+const _ = require('lodash');
```
Comment (lidord-wix) [inline eslint-rules/lib/utils/importUtils.js:1]: This is a really hard-to-understand file..
I think adding a description to the functions can be helpful here :)

**Review by lidord-wix — COMMENTED** [review]
Looks awesome!
I left you some comments.
For the next time - consider dividing big PRs as this one into smaller PRs, just to make it easier to review

File: eslint-rules/lib/utils/deprecationsUtils.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+  return obj;
+}
+
+function organizeDeprecations(deprecations, organizationType, defaultSource) {
+  if (!deprecations) {
+    return {};
+  }
+
+  switch (organizationType) {
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply eslint-rules/lib/utils/deprecationsUtils.js:24]: I thought there might be more types, but you are right.
Removed `OrganizationType`, so even an `if` is not needed.

File: eslint-rules/lib/rules/prop-value-shape-deprecation.js [hunk 3]
```diff
@@ -1,5 +1,5 @@
 const _ = require('lodash');
-const utils = require('../utils');
+const utils = require('../utils_old');
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply eslint-rules/lib/rules/prop-value-shape-deprecation.js:2]: I did, moved all the ones I needed, you'll use the new utils in a future task for this rule and you'll remove it then ;)

File: eslint-rules/lib/rules/component-prop-deprecation.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    function checkPropDeprecation(node, fixNode, propName, deprecatedPropList, componentName) {
+      const deprecatedProp = _.find(deprecatedPropList, {prop: propName});
+      if (deprecatedProp) {
+        const {prop, message, fix} = deprecatedProp;
+        reportDeprecatedProps({node, name: componentName, prop, message, fixNode, fix});
+      }
+    }
+
+    function deprecationCheck(node) {
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply eslint-rules/lib/rules/component-prop-deprecation.js:65]: Touché :)
Fixed.

File: eslint-rules/lib/rules/component-prop-deprecation.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      });
+    }
+
+    return {
+      ImportDeclaration: node => addToImports(node, imports),
+      VariableDeclarator: node => addToImports(node, imports),
+      JSXOpeningElement: node => deprecationCheck(node)
+    };
+  }
+};
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply eslint-rules/lib/rules/component-prop-deprecation.js:112]: True: https://jira.wixpress.com/browse/WOAUILIB-1658

File: eslint-rules/lib/rules/component-prop-deprecation.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+const {
+  organizeDeprecations,
+  OrganizationType,
+  addToImports,
+  getComponentLocalName,
+  getComponentName,
+  findValueNodeOfIdentifier
+} = utils;
+
+const MAP_SCHEMA = {
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply eslint-rules/lib/rules/component-prop-deprecation.js:13]: Lets talk about this, please ping me

File: eslint-rules/lib/rules/typography-deprecation.js [hunk 7]
```diff
@@ -1,25 +1,35 @@
 const _ = require('lodash');
 const utils = require('../utils');
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply eslint-rules/lib/rules/typography-deprecation.js:2]: I guess so, not sure why I did that 🤔

File: eslint-rules/lib/utils/debugUtils.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
+// stringify with no circular error
+function stringify(object) {
+  if (object && typeof object === 'object') {
+    object = copyWithoutCircularReferences([object], object);
+  }
+  return JSON.stringify(object);
+
+  function copyWithoutCircularReferences(references, object) {
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply eslint-rules/lib/utils/debugUtils.js:8]: Thanks, fixed

File: eslint-rules/lib/utils/debugUtils.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    const cleanObject = {};
+    Object.keys(object).forEach(key => {
+      const value = object[key];
+      if (value && typeof value === 'object') {
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply eslint-rules/lib/utils/debugUtils.js:12]: I think you are right, but TBH it's a copy-pasted function and I don't want to change it without testing, I'll add a TODO, and hopefully someone will test it :)

File: eslint-rules/lib/utils/importUtils.js (hunk 1 above)
Comment (M-i-k-e-l) [reply eslint-rules/lib/utils/importUtils.js:1]: I kind of think the method names and the two comments in the exported function are enough, which methods' names do you find confusing?

Comment (M-i-k-e-l) [reply eslint-rules/lib/utils/importUtils.js:1]: And generally I find all lint related code hard-to-understand : ))

Comment (M-i-k-e-l) [reply eslint-rules/lib/utils/importUtils.js:1]: BTW, it seems like you've read most if not all the code... Well done, I think... : ))

Comment (lidord-wix) [reply eslint-rules/lib/utils/importUtils.js:1]: I'm just not sure what the `addToImports` util does..

File: eslint-rules/lib/utils/debugUtils.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (lidord-wix) [reply eslint-rules/lib/utils/debugUtils.js]: `typeof undefined` is `undefined`, so you can safely remove it :)

File: eslint-rules/lib/utils/importUtils.js (hunk 1 above)
Comment (lidord-wix) [reply eslint-rules/lib/utils/importUtils.js:1]: I'm just not sure what the `addToImports` util does..

File: eslint-rules/lib/rules/no-hard-coded-color.js [hunk 10]
```diff
@@ -1,27 +1,24 @@
...
     messages: {
-      uiLib: 'Use UILib colors instead of hardcoded colors.',
+      uiLib: 'Use UILib colors instead of hardcoded colors.'
     },
     fixable: 'whitespace',
-    schema: [
-      MAP_SCHEMA,
-      MAP_SCHEMA,
-    ],
+    schema: [MAP_SCHEMA, MAP_SCHEMA]
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply eslint-rules/lib/rules/no-hard-coded-color.js]: Removed

File: eslint-rules/lib/rules/component-prop-deprecation.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply eslint-rules/lib/rules/component-prop-deprecation.js:11]: Added, however I did not manage to make the tests fail on it

File: eslint-rules/lib/utils/importUtils.js (hunk 1 above)
Comment (M-i-k-e-l) [reply eslint-rules/lib/utils/importUtils.js:1]: Added docs

**Review by lidord-wix — APPROVED** [review]

**Pull request merged**